File: services/api-gateway/src/handlers/provider/toggle-availability.ts
import type { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { AppError, ErrorCode, ProviderStatus } from '@towcommand/core';
import { GeoCache } from '@towcommand/cache';
import { ProviderRepository } from '@towcommand/db';
import { publishEvent, EVENT_CATALOG } from '@towcommand/events';
import { handleError, successResponse } from '../../middleware/error-handler';

const geoCache = new GeoCache();
const providerRepo = new ProviderRepository();

export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
  try {
    const providerId = event.requestContext.authorizer?.providerId as string;
    const body = JSON.parse(event.body ?? '{}');
    const goOnline = body.online === true;

    const provider = await providerRepo.getById(providerId);

    if (!provider) {
      throw AppError.notFound('Provider');
    }

    if (provider.status !== ProviderStatus.ACTIVE) {
      throw AppError.badRequest(
        ErrorCode.PROVIDER_NOT_VERIFIED,
        'Only verified providers can go online',
      );
    }

    await providerRepo.update(providerId, { isOnline: goOnline });

    if (goOnline) {
      // Add to geo index so they appear in nearby searches
      if (provider.currentLat && provider.currentLng) {
        await geoCache.updateProviderLocation(
          providerId, provider.currentLat, provider.currentLng,
          provider.serviceAreas[0] ?? 'NCR',
        );
      }

      await publishEvent(
        EVENT_CATALOG.provider.source,
        EVENT_CATALOG.provider.events.ProviderOnline,
        {
          providerId,
          lat: provider.currentLat,
          lng: provider.currentLng,
          serviceArea: provider.serviceAreas[0] ?? 'NCR',
        },
        { userId: providerId, userType: 'provider' },
      );
    } else {
      // Remove from geo index
      await geoCache.removeProvider(providerId, provider.serviceAreas[0] ?? 'NCR');

      await publishEvent(
        EVENT_CATALOG.provider.source,
        EVENT_CATALOG.provider.events.ProviderOffline,
        { providerId },
        { userId: providerId, userType: 'provider' },
      );
    }

    return successResponse({ providerId, online: goOnline });
  } catch (error) {
    return handleError(error);
  }
}

File: internal/usecase/provider/toggle_availability.go
package provider

import (
	"context"
	"log/slog"

	"github.com/davidlramirez95/towcommand/internal/domain/errors"
	"github.com/davidlramirez95/towcommand/internal/domain/event"
	provdomain "github.com/davidlramirez95/towcommand/internal/domain/provider"
	"github.com/davidlramirez95/towcommand/internal/usecase/port"
)

// ToggleAvailabilityInput carries the validated fields for toggling availability.
type ToggleAvailabilityInput struct {
	ProviderID string `json:"providerId" validate:"required"`
	Online     bool   `json:"online"`
}

// ToggleAvailabilityOutput is the response after toggling availability.
type ToggleAvailabilityOutput struct {
	ProviderID string `json:"providerId"`
	Online     bool   `json:"online"`
}

// ToggleAvailabilityUseCase orchestrates provider availability toggling.
type ToggleAvailabilityUseCase struct {
	finder    port.ProviderFinder
	updater   port.ProviderAvailabilityUpdater
	geo       port.GeoCache
	publisher port.EventPublisher
	logger    *slog.Logger
}

// NewToggleAvailabilityUseCase creates a new ToggleAvailabilityUseCase.
func NewToggleAvailabilityUseCase(
	finder port.ProviderFinder,
	updater port.ProviderAvailabilityUpdater,
	geo port.GeoCache,
	publisher port.EventPublisher,
	logger *slog.Logger,
) *ToggleAvailabilityUseCase {
	return &ToggleAvailabilityUseCase{
		finder: finder, updater: updater, geo: geo, publisher: publisher, logger: logger,
	}
}

// Execute toggles a provider's online status. Only active providers may go online.
func (uc *ToggleAvailabilityUseCase) Execute(ctx context.Context, input ToggleAvailabilityInput) (*ToggleAvailabilityOutput, error) {
	p, err := uc.finder.FindByID(ctx, input.ProviderID)
	if err != nil {
		return nil, errors.NewInternalError("failed to find provider").WithCause(err)
	}
	if p == nil {
		return nil, errors.NewNotFoundError("Provider", input.ProviderID)
	}

	if input.Online && p.Status != provdomain.ProviderStatusActive {
		return nil, errors.NewValidationError("only verified providers can go online")
	}

	if err := uc.updater.UpdateAvailability(ctx, input.ProviderID, input.Online); err != nil {
		return nil, errors.NewInternalError("failed to update availability").WithCause(err)
	}

	if input.Online {
		if p.CurrentLat != nil && p.CurrentLng != nil {
			if err := uc.geo.AddProviderLocation(ctx, input.ProviderID, *p.CurrentLat, *p.CurrentLng); err != nil {
				uc.logger.WarnContext(ctx, "failed to add provider to geo index", "error", err)
			}
		}

		if err := uc.publisher.Publish(ctx, event.SourceProvider, event.ProviderOnline, map[string]any{
			"providerId": input.ProviderID,
			"lat":        p.CurrentLat,
			"lng":        p.CurrentLng,
		}, &port.Actor{UserID: input.ProviderID, UserType: "provider"}); err != nil {
			uc.logger.WarnContext(ctx, "failed to publish ProviderOnline event", "error", err)
		}
	} else {
		if err := uc.geo.RemoveProvider(ctx, input.ProviderID); err != nil {
			uc.logger.WarnContext(ctx, "failed to remove provider from geo index", "error", err)
		}

		if err := uc.publisher.Publish(ctx, event.SourceProvider, event.ProviderOffline, map[string]any{
			"providerId": input.ProviderID,
		}, &port.Actor{UserID: input.ProviderID, UserType: "provider"}); err != nil {
			uc.logger.WarnContext(ctx, "failed to publish ProviderOffline event", "error", err)
		}
	}

	return &ToggleAvailabilityOutput{
		ProviderID: input.ProviderID,
		Online:     input.Online,
	}, nil
}
